### PR TITLE
Add basic support for flat categories

### LINKED ISSUE #4153: Create a "flat" method on the categories controller
Create a flat method on the categories controller that will display the current category's immediate children as a single flat, paged list. Even though this is a flat list, the categories should still be set on the "CategoryTree" key on the dat array.

- [ ] When viewing a category with display as = "flat" then this endpoint should be called. In this way you don't actually browse to the flat endpoint, but rather it gets called internally when a category is supposed to display this way.
- [ ] The endpoint should respect the tree/list config setting.
- [ ] When viewing a regular category tree/list you should not see the children of a category if it has a display as of "flat" or "categories". The user should have to click in to both of these.

### PR DESCRIPTION
**Pull request by initvector**
Closes #4153 

- [x] Display as "Flat" shout show up in the add/edit category page.
- [x] When viewing a flat category headings should be filtered out.
- [x] You should not be allowed to choose a display as type of heading underneath a flat category. In the case where one sneaks in because of the above then that's when it gets filtered out.
- [x] The view should treat categories as depth 1. Right now they are adding Depth2, Depth-2 CSS classes if the category is at that global depth.

### CODE REVIEW

**Comment by initvector** [comment]
This branch has now been rebased.

**Comment by tburry** [comment]
I unchecked one item because you need to make an additional server check and not just filter the UI.

**Comment by initvector** [comment]
Added simple controller-based verification of new "Display As" value when editing a category

File: applications/vanilla/views/categories/flat_all.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+$MaxDisplayDepth = c('Vanilla.Categories.MaxDisplayDepth') + $this->data('Category.Depth', 0);
+$ChildCategories = '';
+
+$CategoryTree = $this->data('CategoryTree');
+$Categories = CategoryModel::flattenTree($CategoryTree);
+$this->EventArguments['NumRows'] = $Categories;
+
+echo '<ul class="DataList CategoryList'.($DoHeadings ? ' CategoryListWithHeadings' : '').'">';
+foreach ($Categories as $CategoryRow) {
+    $Category = (object)$CategoryRow;
```
Comment (tburry) [inline applications/vanilla/views/categories/flat_all.php]: I'm hoping you can try this with WriteListItem() so it's more of a simple view like the table.

**Comment by initvector** [comment]
I added usage of `writeListItem` to the flat_all view.  Doing this made a lot of the code copied from [applications/vanilla/views/categories/all.php](https://github.com/vanilla/vanilla/blob/master/applications/vanilla/views/categories/all.php) redundant.  The view has been trimmed down and its formatting has been slightly updated.

**Pull request merged**